File: CPP/No757.cc
/**
 * Created by Xiaozhong on 2020/12/4.
 * Copyright (c) 2020/12/4 Xiaozhong. All rights reserved.
 */
#include <string>
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int intersectionSizeTwo(vector<vector<int>> &intervals) {
        sort(intervals.begin(), intervals.end(), [](const vector<int> &lhs, const vector<int> &rhs) {
            return lhs[0] != rhs[0] ? lhs[0] < rhs[0] : lhs[1] > rhs[1];
        });
        vector<int> todo(intervals.size(), 2);
        int ans = 0, t = intervals.size();
        while (--t >= 0) {
            int s = intervals[t][0];
            int e = intervals[t][1];
            int m = todo[t];
            for (int p = s; p < s + m; ++p) {
                for (int i = 0; i <= t; ++i) {
                    if (todo[i] > 0 && p <= intervals[i][1])
                        todo[i]--;
                }
                ans++;
            }
        }
        return ans;
    }
};

File: Java/src/No757.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Java
 *
 * @Description: TODO
 * @Author: cherry
 * @Create on: 2022/7/22
 **/

public class No757 {
    public int intersectionSizeTwo(int[][] intervals) {
        int n = intervals.length;
        int res = 0;
        int m = 2;
        Arrays.sort(intervals, (lhs, rhs) -> {
            if (lhs[0] == rhs[0]) return rhs[1] - lhs[1];
            return lhs[0] - rhs[0];
        });

        List<List<Integer>> temp = new ArrayList<>();
        for (int idx = 0; idx < n; ++idx) temp.add(new ArrayList<>());
        for (int idx = n - 1; idx >= 0; --idx) {
            for (int num = intervals[idx][0], size = temp.get(idx).size(); size < m; ++size, ++num) {
                res++;
                help(intervals, temp, idx - 1, num);
            }
        }
        return res;
    }

    private void help(int[][] intervals, List<List<Integer>> temp, int pos, int num) {
        for (int idx = pos; idx >= 0; --idx) {
            if (intervals[idx][1] < num) break;
            temp.get(idx).add(num);
        }
    }
}
